tests/component/encapsulation.cpp

```diff
@@ -124,3 +125,47 @@ TEST(Encapsulation, hierarchyCircular) {
     a_parent = parent->serialise(libcellml::Format::XML);
     EXPECT_EQ(e_parent_2, a_parent);
 }
+
+TEST(Encapsulation, hierarchyWaterfallAndParse) {
+    const std::string e =
+            "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+            "<model xmlns=\"http://www.cellml.org/cellml/1.2#\">"
+            "<component name=\"parent_component\"/>"
+            "<component name=\"child1\"/>"
+            "<component name=\"child2\"/>"
+            "<component name=\"child3\"/>"
+            "<encapsulation>"
+            "<component_ref component=\"parent_component\">"
+            "<component_ref component=\"child1\">"
+            "<component_ref component=\"child2\">"
+            "<component_ref component=\"child3\"/>"
+            "</component_ref>"
+            "</component_ref>"
+            "</component_ref>"
+            "</encapsulation>"
+            "</model>";
```

Comment:
Nitpicking: might want to indent here and there to make it easier to read. In that context, in some places (in the code), we use 4-space indentation while in other cases we use 2-space indentation. It would be nice to be consistent (and probably use 2-space indentation).